File: 0090-subsets-ii/0090-subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        def dfs(index, path):
            if index == len(nums):
                result.append(path[:])
                return

            path.append(nums[index])
            dfs(index + 1, path)
            path.pop()

            while index + 1 < len(nums) and nums[index] == nums[index + 1]:
                index += 1

            dfs(index + 1, path)

        nums.sort()
        result = []
        dfs(0, [])

        return result

File: 0090-subsets-ii/0090-subsets-ii.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsetsWithDup = function (nums) {
    nums.sort((a, b) => a - b);
    
    const dfs = function (index, path) {
        if (index === nums.length) {
            result.push([...path]);
            return;
        }

        path.push(nums[index]);
        dfs(index + 1, path);
        path.pop();

        while (index + 1 < nums.length && nums[index] === nums[index + 1]) {
            index++;
        }

        dfs(index + 1, path);
    }

    const result = [];
    dfs(0, []);

    return result;
};